### PR TITLE
sdk/state: reject payment if a coordinated close already accepted

### PR DESCRIPTION
**Pull request by acharb**
**WHAT**
in ProposePayment and ConfirmPayment check to see if there's an accepted coordinated close , and if so don't allow new payments

**WHY**
with a fully signed coordinated close a participant could submit a C_i/D_i near instantly, which invalidates later payments

closes https://github.com/stellar/experimental-payment-channels/issues/107

### CODE REVIEW

File: sdk/state/payment.go [hunk 1]
```diff
@@ -111,6 +119,10 @@ func (c *Channel) validatePayment(ca CloseAgreement) (err error) {
 		ca.Details.ConfirmingSigner.Address() != c.remoteSigner.Address() {
 		return fmt.Errorf("close agreement confirmer does not match a local or remote signer, got: %s", ca.Details.ConfirmingSigner.Address())
 	}
+	if !c.latestAuthorizedCloseAgreement.isEmpty() && c.latestAuthorizedCloseAgreement.Details.ObservationPeriodTime == 0 &&
```
Comment (acharb) [inline sdk/state/payment.go:122]: I was considering adding a helper method like `isAuthorized()` instead for this check, but thought using `!isEmpty()` is good enough

File: sdk/state/payment.go [hunk 2]
```diff
@@ -54,6 +54,13 @@ func (c *Channel) ProposePayment(amount int64) (CloseAgreement, error) {
 	if amount <= 0 {
 		return CloseAgreement{}, errors.New("payment amount must be greater than 0")
 	}
+
+	// If a coordinated close has been accepted already, error.
```
Comment (leighmcculloch) [inline sdk/state/payment.go:58]: ❗ We should also check if we have proposed a coordinated close by checking if the `latestUnauthorizedCloseAgreement` has zeroes too. If we've proposed a coordinated close we shouldn't ever propose or sign any new close agreements, even if we never get a response back from the other participant. The reason being that they could sign and submit that close at anytime, even if we've both agreed to new close agreements after it.

File: sdk/state/payment.go (hunk 1 above)
Comment (leighmcculloch) [reply sdk/state/payment.go:122]: ❗ Same comment as above, we need to check the unauthorized as well.

**Review by leighmcculloch — COMMENTED** [review]
One more thing that needs adding, otherwise this looks great.

File: sdk/state/payment.go (hunk 2 above)
Comment (acharb) [reply sdk/state/payment.go:57]: 👍 added here: https://github.com/stellar/experimental-payment-channels/pull/197/commits/e342078733285002ecf2f183ae7586a23744d240

File: sdk/state/payment.go [hunk 3]
```diff
@@ -110,16 +110,16 @@ var ErrUnderfunded = fmt.Errorf("account is underfunded to make payment")
...
 	// If a coordinated close has been proposed by this channel already, error.
 	if !c.latestUnauthorizedCloseAgreement.isEmpty() && c.latestUnauthorizedCloseAgreement.Details.ObservationPeriodTime == 0 &&
 		c.latestUnauthorizedCloseAgreement.Details.ObservationPeriodLedgerGap == 0 {
 		return fmt.Errorf("cannot propose payment after proposing a coordinated close")
 	}
+	// If a coordinated close has been accepted already, error.
+	if !c.latestAuthorizedCloseAgreement.isEmpty() && c.latestAuthorizedCloseAgreement.Details.ObservationPeriodTime == 0 &&
+		c.latestAuthorizedCloseAgreement.Details.ObservationPeriodLedgerGap == 0 {
+		return fmt.Errorf("cannot confirm payment after an accepted coordinated close")
+	}
```
Comment (leighmcculloch) [inline sdk/state/payment.go]: Why did these need to be reordered? The commit is named "fix test", but I can't see how reordering these would result in changed behavior.

**Review by leighmcculloch — APPROVED** [review]
🎉 Looks great. I have one question, but looks great regardless.

File: sdk/state/payment.go (hunk 3 above)
Comment (acharb) [reply sdk/state/payment.go]: technically it was already impossible to propose a new payment if you proposed a coordinated close and it's pending
1. if you propose a payment with `O` period of not zero to match the `latestAuthorized` details, you'll fail the compare against `latestUnauthorized` check
2. if you propose a payment with `O` period of zero to match the new `latestUnauthorized` details, you'll fail the compare against `latestAuthorized` check

so to make the check explicit it has to go first 😄

**Pull request merged**